ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -305,16 +310,18 @@ struct vk_buffer_struct {
     void * ptr;
     size_t size = 0;
 
-    vk_device device;
+    vk_device_ref device_ref;
 
     ~vk_buffer_struct() {
-        if (size == 0) {
+        if (size == 0 || device_ref.expired()) {
```

Comment:
If we get here with an expired pointer, does that indicate that we failed to free the VkBuffer/VkDeviceMemory? Is there any way to avoid that?

Reply:
This case happens in the `vk_device_struct` deconstructor, there is still the sync_staging buffer around. But at that point the weak pointer is invalid, so I added a `ggml_vk_destroy_buffer` overload that uses the `vk::Device` directly and added this check. There might be a better way to handle that.